### PR TITLE
drivers: add atmel_saic driver

### PR DESCRIPTION
**Pull request by clementleger**
Add a driver for sama5d2 interrupt handler and initialize it in sama5d2 platform.

### CODE REVIEW

File: core/drivers/atmel_saic.c [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+	int it_grp = it / 32;
+	int it_off = it % 32;
+
+	if (it >= saic.nr_irqs)
+		panic();
+
+	return saic.external[it_grp] & BIT32(it_off);
+}
+
+static uint32_t saic_get_src_type(int dt_level, int it, uint32_t *src_type)
```
Comment (jenswi-linaro) [inline core/drivers/atmel_saic.c:78]: It looks like you should probably use `int` instead as return type.

File: core/drivers/atmel_saic.c [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+	}
+
+	return 0;
+}
+
+static void saic_add(struct itr_chip *chip __unused, size_t it,
+		     uint32_t type, uint32_t prio)
+{
+	uint32_t src_type = AT91_AIC_SMR_HIGH_LEVEL;
+	uint32_t res = TEE_ERROR_GENERIC;
```
Comment (jenswi-linaro) [inline core/drivers/atmel_saic.c:110]: This has nothing to do with the TEE_ERRROR_* codes.

File: core/drivers/atmel_saic.c [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+	io_write32(sfr_base + AT91_SFR_AICREDIR, aicredir_val);
+}
+
+static void saic_init_external(const void *fdt, int node)
+{
+	int i = 0;
+	int len = 0;
+	int it_grp = 0;
+	int it_off = 0;
+	size_t it = 0xFFFFFFFF;
```
Comment (jenswi-linaro) [inline core/drivers/atmel_saic.c:223]: This looks like a magic number.

File: core/drivers/atmel_saic.c [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+	external = fdt_getprop(fdt, node, "atmel,external-irqs", &len);
+	if (!external)
+		return;
+
+	len /= sizeof(uint32_t);
+	for (i = 0; i < len; i++) {
+		it = fdt32_to_cpu(external[i]);
+
+		DMSG("IRQ %d is external", it);
```
Comment (jenswi-linaro) [inline core/drivers/atmel_saic.c:234]: %zu

File: core/drivers/atmel_saic.c (hunk 1 above)
Comment (clementleger) [reply core/drivers/atmel_saic.c:78]: Agreed, to be more coherent with other value, I will switch change this function to return  a real `TEE_Result` and check that in other callers

File: core/drivers/atmel_saic.c (hunk 3 above)
Comment (clementleger) [reply core/drivers/atmel_saic.c:223]: Acked and there is no need for such value.

**Comment by clementleger** [comment]
New version:
- Fixed return type and check for `saic_get_src_type()`
- Remove useless `0xFFFFFFFF` init value
- Simplify check for `saic_init()` in `main_init_gic()` (remove TEE_Result variable)

File: core/arch/arm/plat-sam/main.c [hunk 5]
```diff
@@ -351,3 +353,14 @@ void plat_primary_init_early(void)
 {
 	matrix_init();
 }
+
+void itr_core_handler(void)
+{
+	atmel_saic_it_handle();
+}
+
+void main_init_gic(void)
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-sam/main.c:362]: Hmm, now it's clear that we could have picked a better name for this function. :-)
We should change it in a later PR.

File: core/include/drivers/atmel_saic.h [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+#include <tee_api_types.h>
+
+#define AT91_AIC_WPKEY	0x414943
+
+#define AT91_AIC_SSR		0x0
+#define	AT91_AIC_SSR_ITSEL_MASK	GENMASK_32(6, 0)
+
+#define AT91_AIC_SMR		0x4
+#define	AT91_AIC_SMR_SRC_SHIFT	5
+#define	AT91_AIC_SMR_LEVEL	(0 << AT91_AIC_SMR_SRC_SHIFT)
```
Comment (jenswi-linaro) [inline core/include/drivers/atmel_saic.h:18]: It is generally more safe to use the macro `SHIFT_U32()` for this

File: core/include/drivers/atmel_saic.h [hunk 7]
```diff
@@ -0,0 +1,47 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2021, Microchip
+ */
+
+#ifndef __DRIVERS_ATMEL_SAIC_H
+#define __DRIVERS_ATMEL_SAIC_H
+
+#include <tee_api_types.h>
+
```
Comment (jenswi-linaro) [inline core/include/drivers/atmel_saic.h:10]: Please include <util.h> too.

File: core/arch/arm/plat-sam/main.c (hunk 5 above)
Comment (clementleger) [reply core/arch/arm/plat-sam/main.c:362]: Yep :)

File: core/include/drivers/atmel_saic.h (hunk 6 above)
Comment (clementleger) [reply core/include/drivers/atmel_saic.h:18]: Ack

File: core/include/drivers/atmel_saic.h (hunk 7 above)
Comment (clementleger) [reply core/include/drivers/atmel_saic.h:10]: Ok

File: core/drivers/atmel_saic.c [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+}
+
+static void saic_select_it(size_t it)
+{
+	assert(!(it & ~AT91_AIC_SSR_ITSEL_MASK));
+
+	saic_write_reg(AT91_AIC_SSR, it);
+}
+
+static void saic_configure_it(size_t it, int src_type, int priority)
```
Comment (jenswi-linaro) [inline core/drivers/atmel_saic.c:61]: I would have expected `uint32_t` instead of `int` here.
Any particular reason why `int` is used here?

Comment (clementleger) [reply core/drivers/atmel_saic.c:61]: No, you are right, using `uint32_t` would be better.

Comment (clementleger) [reply core/drivers/atmel_saic.c:61]: I wil lalso change the `it` parameter of `saic_get_src_type()` to use a `size_t`

**Comment by jenswi-linaro** [comment]
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by clementleger** [comment]
New version:
- Added @jenswi-linaro `Acked-by`

**Pull request merged**